### DISCUSSION TITLE
Android Authorities issue in AAB

### DISCUSSION THREAD

**Discussion by mrfrasier** [airsdk/Adobe-Runtime-Support#2981] [Q&A]
Hi all,

I've had an app on Google Play for a decade now with no issues. We switched to AAB and last submission (last year) was fine. Went to update and now when I upload the aab file I get the following error:

"Remove conflicts from the manifest before uploading. The following content provider authorities are in use by other developers: MY.PACKAGE.NAME.provider"

I've checked in the Android manifest and while there are a few provider authorities in there, none match that exactly. There are plenty along the lines of MY.PACKAGE.NAME.firebaseinitprovider, MY.PACKAGE.NAME.sharefileprovider etc, but none that are my package.provider

I'm unsure where this is coming from and where I go from here (indeed, I'm not even sure how, even if that were present, another app could be using it given this has been my package name for ages. Could someone have uploaded an app with a certain provider to try to mess with our own app?)

**Comment by ajwfrost** [comment]
Hi

Well the packages would all have a 'provider' added to them for the camera UI picker, I'm not sure when this was added originally (by Adobe) but it's been there for a while. We had to make a change to add the "exported" flag that is now mandatory, but we set it to false because it's not something that other packages should be able to use!
```
<provider android:authorities="air._our_application_id_.provider"
android:exported="false"
android:grantUriPermissions="true"
android:name="com.adobe.air.CameraUIProvider"
>
```
But then, it's very weird if that "my package name" part is yours, how come it would be in use by other developers?! Unless you'd uploaded this application separately at some point?

I don't know whether it's possible to find out what other developer may have been using your package name, I can't help but think there may be some mis-configuration or glitch within the Android systems that maybe is getting mixed up somehow, as it's very unlikely anyone else would be doing that?!

Lots of other people facing the same issue from what I can see, and no clear answers about how things were solved... e.g.
https://stackoverflow.com/questions/77390849/uploading-app-bundle-to-google-play-console-gives-remove-conflicts-from-the-man

**Reply by mrfrasier** [reply]
Nope, never uploaded the app anywhere else, unless it's looking at old APKs I uploaded and thinking they are somehow different than the AAB.

I asked Google directly about the issue and simply kept getting told I have to change the provider name -- no other real info, but again, if it's a matter of just changing the provider, fine, but WHERE haha.

I know one of the other things I did was update the signing key to use their new signing thingamajig, so I also wonder if it's thinking it's a new app because of a different key being used vs in the past.

I'm used to Apple being the annoying and opaque ones -_-

**Reply by mrfrasier** [reply]
I found this discussion had a detailed answer/suggestion that does bring up the signing certificate as well: https://stackoverflow.com/questions/77217421/when-i-try-to-upload-my-app-bundle-in-google-play-console-i-get-the-error-the

**Reply by marchbold** [reply]
@mrfrasier if you are using any of our extensions make sure you correctly update the  manifest additions. We have several providers in use across our extensions that require you to change the authority to ensure they don't conflict with other applications. 

Our documentation lists the required changes for each extension however worth using apm for things like this to ensure the manifest is correctly updated

**Comment by mrfrasier** [comment]
> @mrfrasier if you are using any of our extensions make sure you correctly update the manifest additions. We have several providers in use across our extensions that require you to change the authority to ensure they don't conflict with other applications.
> 
> Our documentation lists the required changes for each extension however worth using apm for things like this to ensure the manifest is correctly updated

Yup, this project is 100% APM and I do generate app-descriptor as well. One thing I DID change though is I used to edit the manifest myself when switching between Android and iOS since my app is older and has the air. prefix on android but not iOS. Now I edit the apm config and run generate. Wonder if something in there is funky

**Reply by marchbold** [reply]
That's very odd, can you ensure you have the latest androidx dependencies in there. Or can you send me your project and project-lock files.

**Reply by mrfrasier** [reply]
Sorry, was away for CES. Will email you the files shortly

**Comment by mrfrasier** [comment]
So I downloaded the current AAB from Google Play and used Android Studio to extract the manifest. Here is the relevant section for CameraUIProvider as @ajwfrost suspected may be the culprit:

```
<provider
            android:name="com.adobe.air.CameraUIProvider"
            android:exported="false"
            android:authorities="air.air.com.myapp.mobile.provider"
            android:grantUriPermissions="true">

            <meta-data
                android:name="android.support.FILE_PROVIDER_PATHS"
                android:resource="@ref/0x7f110003" />
        </provider>
```

When I examine the aab file that is being built now, here's what I get for that same section:

```
<provider
            android:name="com.adobe.air.CameraUIProvider"
            android:authorities="air.com.myapp.mobile.provider"
            android:exported="false"
            android:grantUriPermissions="true" >
            <meta-data
                android:name="android.support.FILE_PROVIDER_PATHS"
                android:resource="@xml/provider_paths" >
            </meta-data>
        </provider>
```

You'll notice the working version as the authorities as **air.air.com.myapp.mobile.provider** while the non-working version has **air.com.myapp.mobile.provider**

So it does seem there was some funkiness with the noflair tag before and now that I'm using apm to manage project ID and generate manifest I'm seeing some mismatch.

Elsewhere in the manifest the package is correct as **air.com.myapp.mobile.provider**, so it seems it was just the camera ui provider section that was getting screwy.

Any way to sort of force that one section to have the right (well, wrong since it has air twice, but right according to Google Play) prefix? It would appear that's added by the build process as that section is not in my app XML prior to compilation. I tried manually adding the provider hoping that the compiler would then skip over it, but the provider just gets added twice

**Reply by mrfrasier** [reply]
OK following up for anyone hitting this and needing a fix

I found [this answer](https://stackoverflow.com/a/61464613/2149555) to a question on stackoverflow about editing manifest files.

For the sake of posterity, I'm going to break down the steps here (some are copy/pasted from original answer)

1. Build the AAB as usual
2. Open the aab file in a zip program and pull out AndroidManifest.xml (it's under base>manifest). This is not a raw xml file, it needs to be decoded
3. Download protoc binary from https://developers.google.com/protocol-buffers/docs/downloads and extract binary to path or working folder.
4. Go [here](https://maven.google.com/web/index.html#com.android.tools.build:aapt2-proto)and download the latest jar by selecting the latest, and under Artifact(s) clicking jar. Once you get it, open it in a zip program and extract "Resources.proto" & "Configuration.proto" to your working directory
5. Run command protoc --decode=aapt.pb.XmlNode Resources.proto < AndroidManifest.xml > output.txt to decode the file.
6. Find the entry you want to edit (still not an XML file, it's in JSON format, so some searching will be in order) and change it.
7. Run protoc --encode=aapt.pb.XmlNode Resources.proto < output.txt > AndroidManifest_new.xml to encode the new version.
8. Rename it to AndroidManifest.xml (unsure if overwriting the original is a wise idea in step 7) Use zip program to copy the new AndroidManifest.xml back into the aab
9. Re-sign the aab so you can upload it (I used this command jarsigner -verbose -sigalg SHA256withRSA -digestalg SHA-256 -keystore .\mykeystore.p12 .\MyApp.aab 1)

It would still be good to have a cleaner way to do this (maybe there is and andrew can chime in), but this works for now.

**Reply by marchbold** [reply]
What's your app id? Do you have the `air.` in your app id in the app descriptor? 

Also there is an easier way than what you are doing.

In your adt.cfg or via AIR SDK Manager, you can enable "Keep Android Studio Output", then after you do your build you'll have an `AndroidStudioProject` folder alongside your aab. You can open this in Android Studio and edit the AndroidManifest.xml directly and then build your project with Android Studio.

**Reply by mrfrasier** [reply]
I do yes, so in this case in the xml and project.apm file, my ID is entered as **air.com.myapp.mobile**. 

As I mentioned, the project is old, so there has been times when NOFLAIR was on, off etc, and it just seems like something got stuck when we moved to apm and the last release got air prefix added twice. Still unsure why some other app out there has my id without 2 air prefixes, but at least this works for now.

Noted on the Android Studio solution, I'll check into that!
